db/pebble/db.go

```diff
@@ -87,6 +164,192 @@ func (d *DB) WithListener(listener db.EventListener) db.DB {
+		Help:      "Tracks the number of level-zero compactions",
+	})
+	d.nonlevel0CompGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "nonlevel0_comp",
+		Help:      "Tracks the number of non level-zero compactions",
+	})
+	d.seekCompGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "seek_comp",
+		Help:      "Tracks the number of table compaction caused by read opt",
+	})
+	d.manualMemAllocGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "manual_mem_alloc",
+		Help:      "Tracks the amount of non-managed memory currently allocated",
+	})
+
+	prometheus.MustRegister(
+		d.compTimeMeter,
+		d.compReadMeter,
+		d.compWriteMeter,
+		d.writeDelayCountMeter,
+		d.writeDelayTimeMeter,
+		d.diskSizeGauge,
+		d.diskWriteMeter,
+		d.memCompGauge,
+		d.level0CompGauge,
+		d.nonlevel0CompGauge,
+		d.seekCompGauge,
+		d.manualMemAllocGauge,
+	)
+
+	return d
+}
+
+func (d *DB) onCompactionBegin(info pebble.CompactionInfo) { //nolint:gocritic // Used by pebble's event listener
+	if d.activeComp == 0 {
+		d.compStartTime = time.Now()
+	}
+	l0 := info.Input[0]
+	if l0.Level == 0 {
+		d.level0Comp.Add(1)
+	} else {
+		d.nonLevel0Comp.Add(1)
+	}
+	d.activeComp++
+}
+
+func (d *DB) onCompactionEnd(info pebble.CompactionInfo) { //nolint:gocritic // Used by pebble's event listener
+	if d.activeComp == 1 {
+		d.compTime.Add(int64(time.Since(d.compStartTime)))
+	} else if d.activeComp == 0 {
+		panic("should not happen")
+	}
+	d.activeComp--
+}
+
+func (d *DB) onWriteStallBegin(b pebble.WriteStallBeginInfo) {
+	d.writeDelayStartTime = time.Now()
+}
+
+func (d *DB) onWriteStallEnd() {
+	d.writeDelayTime.Add(int64(time.Since(d.writeDelayStartTime)))
+}
+
+// collectMetrics periodically retrieves internal pebble counters and reports them to
+// the metrics subsystem.
+func (d *DB) StartMetricsCollection(ctx context.Context, refresh time.Duration) {
+	timer := time.NewTimer(refresh)
+	defer timer.Stop()
+
+	// Create storage and warning log tracer for write delay.
+	var (
+		compTimes        [2]int64
+		writeDelayTimes  [2]int64
+		writeDelayCounts [2]int64
+		compWrites       [2]int64
+		compReads        [2]int64
+
+		nWrites [2]int64
+	)
+
+	// Iterate ad infinitum and collect the stats
+	for i := 1; ; i++ {
+		var (
+			compWrite int64
+			compRead  int64
+			nWrite    int64
+
+			metrics            = d.pebble.Metrics()
+			compTime           = d.compTime.Load()
+			writeDelayCount    = d.writeDelayCount.Load()
+			writeDelayTime     = d.writeDelayTime.Load()
+			nonLevel0CompCount = int64(d.nonLevel0Comp.Load())
+			level0CompCount    = int64(d.level0Comp.Load())
+		)
+		writeDelayTimes[i%2] = writeDelayTime
+		writeDelayCounts[i%2] = writeDelayCount
+		compTimes[i%2] = compTime
+
+		for j := range metrics.Levels {
+			levelMetrics := metrics.Levels[j]
+			nWrite += int64(levelMetrics.BytesCompacted)
+			nWrite += int64(levelMetrics.BytesFlushed)
```

Comment:
Do we not need other fields like BytesMoved or BytesIngested?

Reply:
You raised a good point :)

`BytesMoved` represents the bytes moved during compactions. Since we're already counting BytesCompacted, we don't need to include BytesMoved to avoid double-counting.

`BytesIngested` represents bytes added through ingestion (e.g., bulk loading). We should include this in our nWrite calculation as it represents additional data written to the database.